### PR TITLE
SI-6260 Avoid double-def error with lambdas over value classes

### PR DESCRIPTION
**Pull request by retronym**
Post-erasure of value classs in method signatures to the underlying
type wreaks havoc when the erased signature overlaps with the
generic signature from an overriden method. There just isn't room
for both. But we _really_ need both; callers to the interface method
will be passing boxed values that the bridge needs to unbox and
pass to the specific method that accepts unboxed values.

This most commonly turns up with value classes that erase to
Object that are used as the parameter or the return type of
an anonymous function.

This was thought to have been intractable, unless we chose
a different name for the unboxed, specific method in the
subclass. But that sounds like a big task that would require
call-site rewriting, ala specialization.

But there is an important special case in which we don't need
to rewrite call sites. If the class defining the method is
anonymous, there is actually no need for the unboxed method;
it will _only_ ever be called via the generic method.

I came to this realisation when looking at how Java 8 lambdas
are handled. I was expecting bridge methods, but found none.
The lambda body is placed directly in a method exactly matching
the generic signature.

This commit detects the clash between bridge and target,
and recovers for anonymous classes by mangling the name
of the target method's symbol. This is used as the bytecode
name. The generic bridge forward to that, as before, with
the requisite box/unbox operations.

Review by @gkossakowski

### CODE REVIEW

**Comment by S11001001** [comment]
Nice, works for me, thanks @retronym .

File: test/files/neg/t6260-named.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class C[A](private val a: Any) extends AnyVal
+trait T[A] {
+  def apply(a: A): A
+}
+
+object Test {
+  (x: C[Any]) => {println(s"f($x)"); x} // okay
+  new T[C[Any]] { def apply(a: C[Any]) = a } // okay
+
+  // we can't rename the specific apply methid to avoid the clash
```
Comment (vendethiel) [inline test/files/neg/t6260-named.scala]: typo "method"

**Comment by adriaanm** [comment]
Nice!

**Comment by retronym** [comment]
I just realised that I should remove the bridge flag. Otherwise people java reflecting on the class might be confused.

**Comment by gkossakowski** [comment]
@retronym: it's pretty awesome you've made that realization you describe! I missed the fact that anonymous classes has fewer constraints they need to respect.

Would it make sense to drop the bridge method in delambdafy phase altogether so we follow what Java 8 does? It would mean that you perform the whole unboxing business in `apply` method and never generate bridge method (name mangled or not). WDYT?

**Comment by retronym** [comment]
That would be a good approach in delambdafy. But maybe harder in erasure. I
opted for the delegation approach as the path of least resistance.

**Comment by gkossakowski** [comment]
Ok. I think we could give it a try when we work on making delambdafy the default. How about you add a TODO comment and fix a typo in the test that @Nami-Doc pointed out?

**Comment by gkossakowski** [comment]
@retronym: ping

**Comment by retronym** [comment]
@gkossakowski fixed. I've also removed the BRIDGE flag, and noted an inconsistency in the way `Delambdafy` names the anonymous classes.

**Comment by gkossakowski** [comment]
PLS REBUILD/pr-scala@d6b1e6e

**Comment by gkossakowski** [comment]
LGTM.

**Pull request merged**